File: Two Pointers & Sliding Window/lengthOfLongestSubstring3.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;
        
        char[] ch = s.toCharArray(); 
        int res = 0;
        
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        for (int j = 0, i = 0; j < s.length(); j++) {
            if (map.containsKey(ch[j])) {
                // i cannot go backward, "abba"
                i = Math.max(map.get(ch[j]) + 1, i); 
            } 
            map.put(ch[j], j);   
            res = Math.max(res, j - i + 1);    
        }
        return res;
    }
}

File: Two Pointers & Sliding Window/lengthOfLongestSubstring3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        res = i = 0
        d = {}
        for j in range(len(s)):
            while s[j] in d and d[s[j]] >= i:
                i = d[s[j]] + 1 # remove dup on the left end
            res = max(res, j - i + 1)
            d[s[j]] = j
        return res